File: AI_Visual_Workflow_Builder/backend/models/workflow.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from enum import Enum

class NodeType(str, Enum):
    LLM_ANALYZE = "LLM_ANALYZE"
    FILTER = "FILTER"
    TRANSFORM = "TRANSFORM"
    NOTIFY = "NOTIFY"

class ExecutionStatus(str, Enum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    SUCCESS = "SUCCESS"
    FAILED = "FAILED"

class NodeConfig(BaseModel):
    prompt: Optional[str] = None
    regex: Optional[str] = None
    mapping: Optional[Dict[str, str]] = None
    target: Optional[str] = None

class Node(BaseModel):
    id: str
    type: NodeType
    config: NodeConfig = Field(default_factory=NodeConfig)
    status: ExecutionStatus = ExecutionStatus.PENDING
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None

class Edge(BaseModel):
    id: str
    source: str
    target: str

class Workflow(BaseModel):
    id: str
    name: str
    nodes: List[Node]
    edges: List[Edge]

class ExecutionResult(BaseModel):
    workflow_id: str
    status: ExecutionStatus
    node_results: Dict[str, Any]

File: AI_Visual_Workflow_Builder/frontend/src/types/workflow.ts
export enum NodeType {
    LLM_ANALYZE = 'LLM_ANALYZE',
    FILTER = 'FILTER',
    TRANSFORM = 'TRANSFORM',
    NOTIFY = 'NOTIFY'
}

export enum ExecutionStatus {
    PENDING = 'PENDING',
    RUNNING = 'RUNNING',
    SUCCESS = 'SUCCESS',
    FAILED = 'FAILED'
}

export interface NodeConfig {
    prompt?: string;
    regex?: string;
    mapping?: Record<string, string>;
    target?: string;
}

export interface WorkflowNode {
    id: string;
    type: NodeType | string; // For react flow string compatibility
    position?: { x: number, y: number };
    data: {
        config: NodeConfig;
        status: ExecutionStatus;
        result?: any;
        error?: string;
        label?: string;
    };
}

export interface WorkflowEdge {
    id: string;
    source: string;
    target: string;
}

export interface Workflow {
    id: string;
    name: string;
    nodes: WorkflowNode[];
    edges: WorkflowEdge[];
}

export interface ExecutionResult {
    workflow_id: string;
    status: ExecutionStatus;
    node_results: Record<string, any>;
}
